### PR TITLE
RESTWS-960: Consolidate Location Search Handling

### PR DESCRIPTION
**Pull request by mogoodrich**
## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-960

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java [hunk 1]
```diff
@@ -249,39 +251,54 @@ public void purge(Location location, RequestContext context) throws ResponseExce
...
-		
-		List<Location> locationsByTag = null;
-		List<Location> locationsByQuery = null;
-		
-		if (tagUuid != null) {
-			LocationTag locationTag = locationService.getLocationTagByUuid(tagUuid);
-			locationsByTag = locationService.getLocationsHavingAllTags(Arrays.asList(locationTag));
+
+		List<Location> locationsByTag = Collections.emptyList();
+		List<Location> locationsByQuery = Collections.emptyList();
```
Comment (mseaton) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java:267]: I would set these to null here by default.  That will resolve the issue you highlight below...

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java [hunk 2]
```diff
@@ -249,39 +251,54 @@ public void purge(Location location, RequestContext context) throws ResponseExce
...
+
+		if (tag != null) {
```
Comment (mseaton) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java:269]: I would use StringUtils.isNotBlank(tag), assuming a parameter like `?tag=&q=Test` should not limit by tag.

Comment (mseaton) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java:269]: Then, if this is true, I would initialize the locationsByTag to a new, empty List

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java [hunk 3]
```diff
@@ -249,39 +251,54 @@ public void purge(Location location, RequestContext context) throws ResponseExce
...
+					locationsByTag = Context.getLocationService().getLocationsHavingAllTags(Arrays.asList(locationTag));
+				}
+			} finally {
+				Context.removeProxyPrivilege(VIEW_LOCATIONS); //Not using PrivilegeConstants.VIEW_LOCATIONS which was removed in platform 1.11+
+				Context.removeProxyPrivilege("Get Locations"); //1.11+
+			}
 		}
-		
+
 		if (query != null) {
```
Comment (mseaton) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java:288]: Same here, I would do a StringUtils.isNotBlank

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java [hunk 4]
```diff
@@ -249,39 +251,54 @@ public void purge(Location location, RequestContext context) throws ResponseExce
...
-			locationsByQuery = locationService.getLocations(query);
+			locationsByQuery = Context.getLocationService().getLocations(query);
 		}
-		
-		if (locationsByTag == null) {
+
+		// note that here we check if the actual query parameters are null, not the results
+		// because, for instance, if a tag parameter is specified and no locations are found with that tag, we want to return an empty list,
+		// regardless of the query parameter, and vice versa; note the default value of the results are set to empty lists above
```
Comment (mseaton) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/LocationResource1_8.java:294]: Get rid of this entire comment, after changes made above.  Anything that is not null should be part of the intersection of results.

**Review by mseaton — APPROVED** [review]

**Comment by mogoodrich** [comment]
Thanks @mseaton , merging this in!

**Pull request merged**